Poster-regist#22

nahokyun opened the pull request:
> ## 포스터 등록 기능 구현

#### - MainController에서 poster 이미지 등록을 저장하는 posterJoin이 추가되었습니다.
## MainController.java
``` java
    @PostMapping("/save-poster")
    @ResponseStatus(HttpStatus.OK)
    public String posterJoin(@RequestPart MultipartFile file,@RequestPart Long movieId){
        movieService.addPoster(file,movieId);
        return SUCCESS_SAVE_POSTER.getMessage();
    }
```
#### - poster 등록을 위해 새로운 poster관련 쿼리를 추가하였습니다.
## MovieMapper.xml
```
    <insert id="savePoster" useGeneratedKeys="true" keyProperty="posterId">
        insert into poster(movie_id, save_file_path, updated_at, created_at)
        values (#{movieId}, #{saveFilePath}, current_timestamp , current_timestamp)
    </insert>
```
#### - poster 등록을 처리하는 addPoster 메소드를 비지니스 로직에 추가하였습니다.
## MovieService.java
```java
   public void addPoster(MultipartFile file,Long movieId){

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(FilePath).build();

        try {
            file.transferTo(destination);
            movieMapper.savePoster(poster);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
    }
```
#### - 이미지들이 저장될 경로를 properties에 넣어 fixedPath로 값을 받아오기 위해 @RequiredArgsConstructor를 사용하던것을 제거하고 생성자를 변경하였습니다.
## MovieService.java
```java
    public MovieService(MovieMapper movieMapper,@Value("${fixedPath}") String fixedPath) {
        this.movieMapper = movieMapper;
        this.fixedPath = fixedPath;
    }
```

nahokyun committed:
Poster-regist#22

MainController에서 poster 이미지 등록을 저장하는 posterJoin이 추가되었습니다.

poster 등록을 위해 새로운 poster관련 쿼리를 추가하였습니다.

Jeonminje commented on src/main/java/study/movieservice/controller/MainController.java:
```diff
@@ -16,8 +18,14 @@ public class MainController {
 
     @PostMapping("/save-movies")
     @ResponseStatus(HttpStatus.OK)
-    public String memberJoin(@RequestBody Movie movie) {
+    public String movieJoin(@RequestBody Movie movie) {
         movieService.addMovie(movie);
         return SUCCESS_SAVE_MOVIE.getMessage();
     }
+    @PostMapping("/save-poster")
+    @ResponseStatus(HttpStatus.OK)
+    public String posterJoin(@RequestPart MultipartFile file,@RequestPart Long movieId){
+        movieService.addPoster(file,movieId);
+        return SUCCESS_SAVE_POSTER.getMessage();
+    }
```
영화 정보를 등록할 때, 버튼하나로 포스터와 영화 정보를 한번에 한 컨트롤러에서 받는게 좋다고 생각하는데 어떻게 생각하실까요??

Jeonminje commented on src/main/java/study/movieservice/controller/MainController.java:
```diff
@@ -16,8 +18,14 @@ public class MainController {
 
     @PostMapping("/save-movies")
     @ResponseStatus(HttpStatus.OK)
-    public String memberJoin(@RequestBody Movie movie) {
+    public String movieJoin(@RequestBody Movie movie) {
         movieService.addMovie(movie);
         return SUCCESS_SAVE_MOVIE.getMessage();
     }
+    @PostMapping("/save-poster")
+    @ResponseStatus(HttpStatus.OK)
+    public String posterJoin(@RequestPart MultipartFile file,@RequestPart Long movieId){
+        movieService.addPoster(file,movieId);
+        return SUCCESS_SAVE_POSTER.getMessage();
+    }
```
영화와 포스터를 한 컨트롤러에서 한번의 요청으로 한번에 받는 것은 어떻게 생각하실까요??

nahokyun replied:
영화와 포스터를 한번의 요청으로 한번에 받는경우는 추후 해당 영화에 포스터만 추가되는 경우에는 불필요하게 Movie 객체또한 전달되어야하므로 두 경우를 분리하는것이 더 낫다고 생각했습니다!

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -22,4 +33,22 @@ public void addMovie(Movie inputMovie){
         movieMapper.save(movie);
     }
+
+    public void addPoster(MultipartFile file,Long movieId){
+
+        String originalFileName=file.getOriginalFilename();
+        String FilePath=fixedPath+originalFileName;
+        File destination=new File(FilePath);
+
+        Poster poster = Poster.builder()
+                .movieId(movieId)
+                .saveFilePath(FilePath).build();
+
+        try {
+            file.transferTo(destination);
+            movieMapper.savePoster(poster);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
+        }
+    }
```
포스터 추가라는 관심사와 파일 입출력이라는 관심사를 분리시키면 더 좋을 것 같네요 :)

nahokyun replied:
수정해보겠습니다!

nahokyun committed:
Poster-regist#22

addPoster에서 파일입출력에 해당하는 부분을 분리하여 saveFile이라는 메소드를 추가하였습니다.

nahokyun commented:
> ### 변경된점
- addPoster내의 관심사 분리를 진행하였습니다. (poster 추가와 파일 입출력)
- 저장 경로를 return해주는 saveFile 메소드를 추가하였습니다.
## MovieService.java
### 변경 전 
```java
public void addPoster(MultipartFile file,Long movieId){

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(FilePath).build();

        try {
            file.transferTo(destination);
            movieMapper.savePoster(poster);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
    }
```
### 변경 후
```java
public void addPoster(MultipartFile file,Long movieId){

        String FilePath=saveFile(file);

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(FilePath).build();
        movieMapper.savePoster(poster);
    }
```
```java
    public String saveFile(MultipartFile file){

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        try {
            file.transferTo(destination);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
        return FilePath;
    }

```

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -1,25 +1,52 @@
+        movieMapper.savePoster(poster);
+    }
+
+    public String saveFile(MultipartFile file){
+
+        String originalFileName=file.getOriginalFilename();
+        String FilePath=fixedPath+originalFileName;
+        File destination=new File(FilePath);
+
+        try {
+            file.transferTo(destination);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
+        }
+        return FilePath;
+    }
```
이 메소드는 MovieService에만 국한된 것은 아닌 것 같습니다. 또 파일 입출력을 현재는 로컬에서 처리하지만 나중에는 S3와도 연동할 수 있으니 추상화를 적용할 수도 있겠네요 :)

nahokyun replied:
한정적인 사용일거라 생각하여 MovieService에 구현했지만 확장성을 고려한다면 따로 class를 빼서 사용하는것이 더 좋을것 같네요. 추상화를 적용하면서 수정해보겠습니다.

sunwootest commented on src/main/java/study/movieservice/repository/MovieMapper.java:
```diff
@@ -2,10 +2,12 @@
 import study.movieservice.domain.movie.Movie;
+import study.movieservice.domain.movie.Poster;
 
 @Mapper
 public interface MovieMapper {
 
     void save(Movie movie);
 
+    void savePoster(Poster poster);
```
Movie와 Poster는 서로 다른 테이블 아닌가요? 같은 Mapper에서 create를 만드셨네요 :)

nahokyun replied:
Poster가 아무래도 Movie에 종속적인 테이블이라고 생각해서 같은 Mapper에 넣었는데 잘못 생각한것 같습니다. 분리하도록 하겠습니다!

nahokyun committed:
Poster-regist#22

파일 입출력 부분을 추상화하여 FileIO 인터페이스를 만들었고, 현재는 FileIOLocal을 적용하고 추후 S3를 적용하면 쉽게 바꿀수 있도록 설계하였습니다.

PosterMapper부분을 추가하여 MovieMapper와 분리하였습니다.

nahokyun commented:
> ### 변경된점
- FileIO인터페이스를 추가하여 local에 저장할때와 S3와 연동할때 손쉽게 바꿀수있도록 추상화를 적용하였습니다.
- MovieMapper에서 PosterMapper에 해당하는 부분을 분리하였습니다.
## FileIO.java
###  
```java

public interface FileIO {
    String saveFile(MultipartFile file);
}
```
## FileIOLocal.java
###  
```java
@Component
public class FileIOLocal implements FileIO {

    private final String fixedPath;

    public FileIOLocal(@Value("${fixedPath}") String fixedPath) {
        this.fixedPath = fixedPath;
    }

    @Override
    public String saveFile(MultipartFile file) {

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        try {
            file.transferTo(destination);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
        return FilePath;
    }
}
```
## PosterMapper.java 추가
### 
```java
@Mapper
public interface PosterMapper {

    void savePoster(Poster poster);
}


```
## MovieService.java의 생성자 변경
### 변경 전
```java
    private final MovieMapper movieMapper;
    private final String fixedPath;

    public MovieService(MovieMapper movieMapper,@Value("${fixedPath}") String fixedPath) {
        this.movieMapper = movieMapper;
        this.fixedPath = fixedPath;
    }

```
### 변경 후
```java
    private final MovieMapper movieMapper;
    private final PosterMapper posterMapper;
    private final FileIO fileIO;

    public MovieService(MovieMapper movieMapper, PosterMapper posterMapper, FileIOLocal fileIOLocal) {
        this.movieMapper = movieMapper;
        this.posterMapper=posterMapper;
        this.fileIO=fileIOLocal;
    }
```

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -1,25 +1,38 @@
-@RequiredArgsConstructor
 public class MovieService {
 
     private final MovieMapper movieMapper;
+    private final PosterMapper posterMapper;
```
개발을 100프로 객체지향적으로 설계할 수는 없습니다. Poster기능 자체가 별로 없더라도 PosterService를 만들고 MovieService클래스에 PosterService를 주입하는 것이 더 나은 방향같습니다. 이 부분은 tradeoff를 고려해서 개발 진행하시면 될 것 같습니다 :)

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -1,25 +1,38 @@
-                .runningTime(inputMovie.getRunningTime())
-                .supervisor(inputMovie.getSupervisor())
-                .openingDate(inputMovie.getOpeningDate()).build();
+    public MovieService(MovieMapper movieMapper, PosterMapper posterMapper, FileIOLocal fileIOLocal) {
+        this.movieMapper = movieMapper;
+        this.posterMapper=posterMapper;
+        this.fileIO=fileIOLocal;
+    }
```
롬복을 사용하지않고 생성자를 만든 이유가 따로 있으신가요?

nahokyun replied:
파일 입출력부분을 분리하기전 properties정보를 사용하기 위해서 생성자를 따로 만들었었는데 역할은 롬복을 사용했을때와 다르지 않기때문에 새롭게 추가된 FileIO 부분만 추가해서 사용했습니다! 다른 경우와의 통일성을 위해 롬복을 사용하는것으로 하는것이 더 나을까요?

sunwootest replied:
properties값을 넣어줄게 아니라면 롬복으로 처리하는게 나은 것 같습니다 :)

sunwootest commented on src/main/java/study/movieservice/service/fileIO/FileIOLocal.java:
```diff
@@ -0,0 +1,33 @@
+import java.io.IOException;
+import static study.movieservice.domain.ExceptionMessageConst.FAILED_FILE_RECEIVE;
+
+@Component
+public class FileIOLocal implements FileIO {
+
+    private final String fixedPath;
```
이 클래스도 비즈니스 로직에 해당되기 때문에 서비스 어노테이션을 사용해도 될 것 같습니다 :)

sunwootest commented on src/main/java/study/movieservice/service/fileIO/FileIO.java:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.web.multipart.MultipartFile;
+
+public interface FileIO {
+    String saveFile(MultipartFile file);
```
upload라는 메서드명을 제안해봅니다.

sunwootest commented on src/main/java/study/movieservice/service/fileIO/FileIOLocal.java:
```diff
@@ -0,0 +1,33 @@
+
+        try {
+            file.transferTo(destination);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
+        }
```
IOException을 여기서 잡아서 처리해주면 추후에 파일입출력을 사용하는 다른 비즈니스 로직에서 예외처리를 할 수 있을까요? 예를들어서 파일입출력이 실패하면 이전의 작업들을 rollback해야하는 경우가 생길 수 있는데 이런 경우도 대비해서 예외를 던지는 것이 나아보입니다.

nahokyun replied:
현재 클래스에서는 throws IOException을 통해 상위레벨로 처리를 미루고 MovieService의 addPoster 메소드 실행시 catch해주는 방식으로 바꾸면 좀 더 유연한 사용이 가능할것 같습니다.

nahokyun commented:
> ### 변경된점
- FileIOLocal.java의 어노테이션을 component에서 Service로 변경하였습니다.
- 예외처리 로직을 변경하였습니다.
- MovieService.java의 생성자 사용을 롬복으로 대체하였습니다.
## FileIOLocal.java
### 변경 전 
```java
@Component
public class FileIOLocal implements FileIO {

    private final String fixedPath;

    public FileIOLocal(@Value("${fixedPath}") String fixedPath) {
        this.fixedPath = fixedPath;
    }

    @Override
    public String saveFile(MultipartFile file) {

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        try {
            file.transferTo(destination);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
        return FilePath;
    }
}
```
### 변경 후 
```java
@Service
public class FileIOLocal implements FileIO {

    private final String fixedPath;

    public FileIOLocal(@Value("${fixedPath}") String fixedPath) {
        this.fixedPath = fixedPath;
    }

    @Override
    public String uploadFile(MultipartFile file) throws IOException {

        String originalFileName = file.getOriginalFilename();
        String FilePath = fixedPath + originalFileName;
        File destination = new File(FilePath);

        file.transferTo(destination);

        return FilePath;
    }
}
```
## MovieService.java
### 변경 전 
```java
public void addPoster(MultipartFile file,Long movieId){

        String FilePath= fileIO.saveFile(file);

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(FilePath).build();
        posterMapper.savePoster(poster);
    }
```
### 변경 후 
```java
public void addPoster(MultipartFile file, Long movieId) {
        try {
            String FilePath = fileIO.uploadFile(file);
            Poster poster = Poster.builder()
                    .movieId(movieId)
                    .saveFilePath(FilePath).build();

            posterMapper.savePoster(poster);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
    }
```

nahokyun committed:
Poster-regist#22

범용적인 메소드이름으로 변경하였습니다.

한정적이게 사용될 수 있는 예외처리를 상위레벨로 처리를 미뤄 유연한 사용을 할수있도록 변경하였습니다.

properties를 사용하지않는 생성자를 롬복으로 대체하였습니다.

nahokyun commented:
> ### 변경된점
- FileIOLocal.java의 어노테이션을 component에서 Service로 변경하였습니다.
- 예외처리 로직을 변경하였습니다.
- MovieService.java의 생성자 사용을 롬복으로 대체하였습니다.
## FileIOLocal.java
### 변경 전 
```java
@Component
public class FileIOLocal implements FileIO {

    private final String fixedPath;

    public FileIOLocal(@Value("${fixedPath}") String fixedPath) {
        this.fixedPath = fixedPath;
    }

    @Override
    public String saveFile(MultipartFile file) {

        String originalFileName=file.getOriginalFilename();
        String FilePath=fixedPath+originalFileName;
        File destination=new File(FilePath);

        try {
            file.transferTo(destination);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
        return FilePath;
    }
}
```
### 변경 후 
```java
@Service
public class FileIOLocal implements FileIO {

    private final String fixedPath;

    public FileIOLocal(@Value("${fixedPath}") String fixedPath) {
        this.fixedPath = fixedPath;
    }

    @Override
    public String uploadFile(MultipartFile file) throws IOException {

        String originalFileName = file.getOriginalFilename();
        String FilePath = fixedPath + originalFileName;
        File destination = new File(FilePath);

        file.transferTo(destination);

        return FilePath;
    }
}
```
## MovieService.java
### 변경 전 
```java
public void addPoster(MultipartFile file,Long movieId){

        String FilePath= fileIO.saveFile(file);

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(FilePath).build();
        posterMapper.savePoster(poster);
    }
```
### 변경 후 
```java
public void addPoster(MultipartFile file, Long movieId) {
        try {
            String FilePath = fileIO.uploadFile(file);
            Poster poster = Poster.builder()
                    .movieId(movieId)
                    .saveFilePath(FilePath).build();

            posterMapper.savePoster(poster);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
    }
```

Jeonminje commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -2,24 +2,37 @@
+            Poster poster = Poster.builder()
+                    .movieId(movieId)
+                    .saveFilePath(FilePath).build();
+
```
이 부분에서 IOException이 발생 하는지 궁금합니다! 발생하지 않는다면 try문 밖으로 빼주는 것은 어떨까요??

nahokyun replied:
String filePath를 try 전에 선언한다면 Poster 저장부분을 try catch 이후로 뺄 수 있을것같습니다.

nahokyun committed:
Poster-regist#22

예외가 발생할 부분만 try-catch문에 남기고 나머지 부분은 이후 처리하는것으로 변경하였습니다.

nahokyun commented:
> ### 변경된점
- 예외가 발생할 부분만 try catch문에 남기도록 변경하였습니다.
## MovieService.java
### 변경 전 
```java
public void addPoster(MultipartFile file, Long movieId) {
        try {
            String FilePath = fileIO.uploadFile(file);
            Poster poster = Poster.builder()
                    .movieId(movieId)
                    .saveFilePath(FilePath).build();

            posterMapper.savePoster(poster);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }
    }
```
### 변경 후 
```java
    public void addPoster(MultipartFile file, Long movieId) {
        String filePath;

        try {
            filePath = fileIO.uploadFile(file);
        } catch (IOException e) {
            throw new IllegalArgumentException(FAILED_FILE_RECEIVE.getMessage());
        }

        Poster poster = Poster.builder()
                .movieId(movieId)
                .saveFilePath(filePath).build();

        posterMapper.savePoster(poster);
    }
```

sunwootest reviewed (approved):
Approve 드립니다 :)

nahokyun committed:
Merge branch 'Movie-regist#18' into Poster-regist#22
